linker/Mono.Linker.Steps/MarkStep.cs

```diff
@@ -1150,18 +1150,52 @@ void ApplyPreserveMethods (MethodDefinition method)
 			MarkMethodCollection (list);
 		}
 
-		protected void MarkFields (TypeDefinition type, bool includeStatic)
+		protected void MarkFields (TypeDefinition type, bool includeStatic, bool markBackingFieldsOnlyIfPropertyMarked = false)
 		{
 			if (!type.HasFields)
 				return;
 
 			foreach (FieldDefinition field in type.Fields) {
 				if (!includeStatic && field.IsStatic)
 					continue;
+
+				if (markBackingFieldsOnlyIfPropertyMarked && field.Name.EndsWith (">k__BackingField")) {
+					var expectedPropertyName = field.Name.Substring (1, field.Name.LastIndexOf ('>') - 1);
+
+					// With mcs we can create the expected property name from the field name itself.
+					var propertyDefinition = type.Properties.FirstOrDefault (p => p.Name == expectedPropertyName);
+
+					// However for assemblies compiled with csc we may not be able to because some system types
+					// use a short name in the backing field name.
+					// For example:
+					// Field Name = <IFoo<int>.Bar>k__BackingField
+					// Property Name = IFoo<System.Int32>.Bar
+					// 
+					// Parsing the string and checking for these system types and converting the names would be ugly and error prone.
+					// Instead, let's search the properties and find the one that makes use of the current backing field
+					if (propertyDefinition == null) {
+						propertyDefinition = SearchPropertiesForMatchingFieldDefinition (field);
```

Comment:
Perhaps it'd be better to maintain only one way of propertyDefinition look up and this one should cover both cases